kernel/src/udp_port_table.rs

```diff
@@ -0,0 +1,249 @@
+// UdpPortSocket allocated.
+static mut PORT_TABLE: [Option<PortEntry>; MAX_NUM_BOUND_PORTS] = [None; MAX_NUM_BOUND_PORTS];
+
+/// The PortQuery trait enables the UdpPortTable to query the userspace bound
+/// ports in the UDP driver. The UDP driver struct implements this trait.
+pub trait PortQuery {
+    fn is_bound(&self, port: u16) -> bool;
+}
+
+/// A UdpPortSocket provides a handle into the bound port table. When binding to
+/// a port, the socket is consumed and stored inside a UdpPortBinding. When
+/// undbinding, the socket is returned and can be used to bind to other ports.
+#[derive(Debug)]
+pub struct UdpPortSocket {
+    idx: usize,
+    port_table: &'static UdpPortTable,
+}
+
+/// The UdpPortTable maintains a reference the port_array, which manages what
+/// ports are bound at any given moment, and user_ports, which provides a
+/// handle to userspace port bindings in the UDP driver.
+pub struct UdpPortTable {
+    port_array: TakeCell<'static, [Option<PortEntry>]>,
+    user_ports: OptionalCell<&'static dyn PortQuery>,
+}
+
+impl fmt::Debug for UdpPortTable {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "[Port Table]")
+    }
+}
+
+impl UdpPortSocket {
+    // important that this function is not public. If it were, capsules could
+    // obtain access to ports bound by other capsules
+    fn new(idx: usize, pt: &'static UdpPortTable) -> UdpPortSocket {
+        UdpPortSocket {
+            idx: idx,
+            port_table: pt,
+        }
+    }
+}
+
+impl Drop for UdpPortSocket {
+    fn drop(&mut self) {
+        self.port_table.destroy_socket(self);
+    }
+}
+
+/// An opaque descriptor that allows the holder to obtain a binding on a port
+/// for receiving UDP packets.
+#[derive(Debug)]
+pub struct UdpReceiverBinding {
+    idx: usize,
+    port: u16,
+}
+
+/// An opaque descriptor that allows the holder to obtain a binding on a port
+/// for sending UDP packets.
+#[derive(Debug)]
+pub struct UdpSenderBinding {
+    idx: usize,
+    port: u16,
+}
+
+impl UdpSenderBinding {
+    fn new(idx: usize, port: u16) -> UdpSenderBinding {
+        UdpSenderBinding {
+            idx: idx,
+            port: port,
+        }
+    }
+
+    pub fn get_port(&self) -> u16 {
+        self.port
+    }
+}
+
+impl UdpReceiverBinding {
+    fn new(idx: usize, port: u16) -> UdpReceiverBinding {
+        UdpReceiverBinding {
+            idx: idx,
+            port: port,
+        }
+    }
+
+    pub fn get_port(&self) -> u16 {
+        self.port
+    }
+}
+
+impl UdpPortTable {
+    // Mark new as unsafe so that the port table is only generated in trusted
+    // code.
+    pub unsafe fn new() -> UdpPortTable {
+        UdpPortTable {
+            port_array: TakeCell::new(&mut PORT_TABLE),
+            user_ports: OptionalCell::empty(),
+        }
+    }
+
+    pub unsafe fn set_user_ports(&self, user_ports_ref: &'static dyn PortQuery) {
```

Comment:
This looks like a case where it shouldn't be unsafe, but should rely on capabilities.